src/MagnumPlugins/IcoImporter/IcoImporter.cpp

```diff
@@ -0,0 +1,256 @@
+    *outByte0 = static_cast<char>(in);
+    *outByte1 = static_cast<char>(in >> 8u);
+    *outByte2 = static_cast<char>(in >> 16u);
+    *outByte3 = static_cast<char>(in >> 24u);
+}
+*/
+
+/*
+struct BitmapInfoHeader {
+    static constexpr std::size_t Size = 40;
+    enum class CompressionMethod : UnsignedInt {
+        BI_RGB = 0,
+        BI_RLE8 = 1,
+        BI_RLE4 = 2,
+        BI_BITFIELDS = 3,
+        BI_JPEG = 4,
+        BI_PNG = 5,
+        BI_ALPHABITFIELDS = 6,
+        BI_CMYK = 11,
+        BI_CMYKRLE8 = 12,
+        BI_CMYKRLE4 = 13
+    };
+
+    UnsignedInt headerSize;
+    Int width;
+    Int height;
+    UnsignedShort colorPlanes; // must be 1
+    UnsignedShort bitsPerPixel; // Typical values are 1, 4, 8, 16, 24 and 32
+    CompressionMethod compressionMethod;
+    UnsignedInt imageDataSize;
+    Int horzontalDensity; // pixel per metre
+    Int verticalDensity; // pixel per metre
+    UnsignedInt colorCount; // the number of colors in the color palette, or 0 to default to 2^n
+    UnsignedInt importantColors; // the number of important colors used, or 0 when every color is important; generally ignored
+};
+static_assert(sizeof(BitmapInfoHeader) == BitmapInfoHeader::Size, "sizeof(BitmapInfoHeader) is incorrect");
+*/
+
+struct IconDir {
+    static constexpr std::size_t Size = 6;
+    UnsignedShort reserved; // must be 0
+    UnsignedShort imageType; // must be either 1 for .ICO, or 2 for .CUR
+    UnsignedShort imageCount;
+};
+static_assert(sizeof(IconDir) == IconDir::Size, "sizeof(IconDir) is incorrect");
+
+struct IconDirEntry {
+    static constexpr std::size_t Size = 16;
+    UnsignedByte imageWidth; // pixels. Value 0 means image width is 256 pixels. :[
+    UnsignedByte imageHeight; // pixels. Value 0 means image width is 256 pixels. :[
+    UnsignedByte colorCount;
+    UnsignedByte reserved; // must be 0
+    union {
+        UnsignedShort colorPlanes;
+        UnsignedShort hotspotX;
+    };
+    union {
+        UnsignedShort bitsPerPixel;
+        UnsignedShort hotspotY;
+    };
+    UnsignedInt imageDataSize;
+    UnsignedInt imageDataOffset;
+};
+static_assert(sizeof(IconDirEntry) == IconDirEntry::Size, "sizeof(IconDirEntry) is incorrect");
+
+}
+
+IcoImporter::IcoImporter() = default;
+
+IcoImporter::IcoImporter(PluginManager::AbstractManager& manager, const std::string& plugin) : AbstractImporter{manager, plugin} {}
+
+IcoImporter::~IcoImporter() = default;
+
+auto IcoImporter::doFeatures() const -> Features { return Feature::OpenData; }
+
+bool IcoImporter::doIsOpened() const { return _isOpened; }
+
+void IcoImporter::doClose() {
+    _isOpened = false;
+    _imageDataArray = nullptr;
+}
+
+namespace {
+    constexpr std::initializer_list<const char> bmpFileHeaderTemplate{'B', 'M', 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+    constexpr std::initializer_list<const char> pngMagicNum{static_cast<char>(0x89), 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
```

Comment:
IIRC `initializer_list` isn't C++11, so this won't work (the CIs will complain). What about a plain `const char pngMagicNum[] { ... }`, and then 

```cpp
std::equal(std::begin(pngMagicNum), std::end(pngMagicNum), imageDataBegin)
```

in the code below?

Reply:
hmm, https://en.cppreference.com/w/cpp/utility/initializer_list says it is c++11

Reply:
Blargh, sorry, forgot a word, it should have said "isn't **constexpr** in C++11".